Add code

dmitry-zhbankov opened the pull request.

dmitry-zhbankov committed:
Add code

msiarko commented:
Don't forget about code formatting

dmitry-zhbankov committed:
Extract function from min/max
Format code

dmitry-zhbankov committed:
Add tests

valerydolyuk commented on test/test.js:
```diff
@@ -0,0 +1,60 @@
+    });
+    describe("function getMaxSubSum()", function () {
+        it("should return 14 when value is [1, 3, 2, 5, 3]", function () {
+            assert.deepStrictEqual(arrLib.getMaxSubSum([1, 3, 2, 5, 3]), 14);
```
Deep comparing here and below is not needed

dmitry-zhbankov replied:
I can't use just assert.equal() on arrays because an array is a reference type in JavaScript, and hence only the reference is compared but I can avoid strict comparison

valerydolyuk commented on arr-lib.js:
```diff
@@ -0,0 +1,110 @@
+        return min_max(arr, "min");
+    };
+
+    let min_max = function (arr, type) {
```
1. It's not great idea to write common method to min and max utils: you increase amount of code lines and you write `switch` statement (using of `switch` usually is bad practice)
2. You may just use methods `Math.max` and `Math.min`

valerydolyuk commented on arr-lib.js:
```diff
@@ -0,0 +1,110 @@
+            case "selection":
+                return sortSelection(arr);
+            default:
+                return sortBubble(arr);
```
I expect to get an error when pass unsupported sorting algorithm (for an instance "shell")

valerydolyuk commented on arr-lib.js:
```diff
@@ -0,0 +1,110 @@
+            swapped = false;
+            for (let j = 0; j < n - i - 1; j++) {
+                if (arr[j] > arr[j + 1]) {
+                    let tmp = arr[j];
```
Swap logic should be moved to a separate common method

valerydolyuk commented on arr-lib.js:
```diff
@@ -0,0 +1,110 @@
+        }
+    };
+
+    context.getMaxSubSum = function (arr) {
```
1. Returns `0` when is called with `[-1]` parameter.
2. Complexity of algorythm might be reduced to O(n^3)

dmitry-zhbankov replied:
The complexity is n^2 \+ 2n where n is a number of `sum += arr[k]` operations. It is less than n^3 for n &gt; 2

valerydolyuk reviewed (changes requested):
Plese, fix comments

dmitry-zhbankov committed:
Fix issues:
Add min/max functions
Extract swap() method
Remove unnecessary strict comparisons
Add tests
